components/patient-data/impl/src/main/java/org/phenotips/data/internal/PatientXWikiConsentManager.java

```diff
@@ -0,0 +1,363 @@
+    {
+        this.refreshSystemConsents();
+    }
+
+    @Override public List<Consent> getSystemConsents()
+    {
+        return systemConsents;
+    }
+
+    @Override public List<Consent> loadConsentsFromPatient(String patientId)
+    {
+        return this.loadConsentsFromPatient(repository.getPatientById(patientId));
+    }
+
+    @Override public List<Consent> loadConsentsFromPatient(Patient patient)
+    {
+        List<Consent> patientConsents = new LinkedList<>();
+        /* List of consent ids a patient has agreed to, read from the database */
+        List<String> xwikiPatientConsents = new LinkedList<>();
+
+        try {
+            DocumentModelBridge patientDocBridge = bridge.getDocument(patient.getDocument());
+            XWikiDocument patientDoc = (XWikiDocument) patientDocBridge;
+            xwikiPatientConsents = readConsentIdsFromPatientDoc(patientDoc);
+        } catch (Exception ex) {
+            this.logger.error(
+                "Could not load patient document {} or read consents. {}", patient.getId(), ex.getMessage());
+        }
+
+        /* Using system consents to determine what consents a patient has agreed to, but not reusing the system consents
+         cache, since those should not have a status. */
+        for (Consent systemConsent : this.systemConsents) {
+            Consent copy = DefaultConsent.copy(systemConsent);
+            if (xwikiPatientConsents.contains(systemConsent.getId())) {
+                copy.setStatus(ConsentStatus.YES);
+            } else {
+                copy.setStatus(ConsentStatus.NO);
+            }
+            patientConsents.add(copy);
+        }
+
+        return patientConsents;
+    }
+
+    private List<String> readConsentIdsFromPatientDoc(XWikiDocument doc)
+    {
+        List<String> ids = new LinkedList<>();
+        BaseObject idsHolder = doc.getXObject(consentIdsHolderReference);
+        if (idsHolder != null) {
+            ids = idsHolder.getListValue(GRANTED);
+        }
+        return ids;
+    }
+
+    @Override public boolean setPatientConsents(Patient patient, Iterable<String> consents)
+    {
+        try {
+            List<Consent> existingConsents = this.selectFromSystem(consents);
+            SaveablePatientConsentHolder holder = this.getPatientConsentHolder(patient);
+            holder.setConsents(convertToIds(existingConsents));
+            holder.save();
+            return true;
+        } catch (Exception ex) {
+            this.logger.error("Could not update consents in patient record {}. {}", patient, ex.getMessage());
+        }
+        return false;
+    }
+
+    /** @return consents that exist in the system and correspond to the given ids */
+    private List<Consent> selectFromSystem(Iterable<String> ids)
+    {
+        List<Consent> existingConsents = new LinkedList<>();
+        for (String id : ids) {
+            for (Consent consent : this.getSystemConsents()) {
```

Comment:
again, you're re-implementing `Collection.contains()`. You should have these functions take `Collection`, not `Iterable` arguments.